components/local-app/pkg/bastion/bastion.go

```diff
@@ -478,7 +478,7 @@ func (b *Bastion) connectTunnelClient(ctx context.Context, ws *Workspace) error
 
 func newTunnelClient(ctx context.Context, ws *Workspace, reconnecting *gitpod.ReconnectingWebsocket) (client *TunnelClient, closed chan struct{}, err error) {
 	logrus.WithField("workspace", ws.WorkspaceID).Info("tunnel: trying to connect ssh client...")
-	err = reconnecting.EnsureConnection(func(conn *gitpod.WebsocketConnection) (bool, error) {
+	err = reconnecting.EnsureConnection(ctx, func(conn *gitpod.WebsocketConnection) (bool, error) {
```

Comment:
a callback should return `true, nil` if ctx is cancelled instead, in such case `EnsureConnection` will close the underlying connection properly and exit `newTunnelCient` with closed err

Reply:
We should also then remove this line then: https://github.com/gitpod-io/gitpod/blob/69a5ff6f73ea5e1cb9b1b42fd27c31baf3d5f2e2/components/local-app/pkg/bastion/bastion.go#L467

Reply:
> Updated: maybe just go func() { <-ctx.Done() ws.Close() } instead in connectTunnelClient

That was already there, but the rc.connect() could never exit.

I reworked ReconnectingWebsocket to use context in a more canonical way which resulted in a cleaner solution and allows use of a connection timeout for e.g.

Reply:
Looks much cleaner ❤️